server/bundler/extensions-loader.js

```diff
@@ -0,0 +1,72 @@
+const fs = require( 'fs' );
+const path = require( 'path' );
+const camelCase = require( 'lodash' ).camelCase;
+const compact = require( 'lodash' ).compact;
+
+function getReducerPath( extensionDir ) {
+	return path.join( 'client', 'extensions', extensionDir, 'state', 'reducer.js' );
+}
+
+function generateReducerRequireString( extensionDir ) {
+	return `'${ camelCase( extensionDir ) }': require( 'extensions/${ extensionDir }/state/reducer' )`;
+}
+
+function generateExtensionsModuleString( reducerRequires ) {
+	return `module.exports = {
+		reducers: function() {
+			return {
+				${ reducerRequires.join( '\n' ) }
+			};
+		}
+	};`;
+}
+
+function getReducersSync( extensionDirs = [] ) {
+	return extensionDirs
+		.filter( ( extensionDir ) => fs.existsSync( getReducerPath( extensionDir ) ) )
+		.map( ( extensionDir ) => generateReducerRequireString( extensionDir ) );
+}
+
+function getExtensionsModuleSync( extensionDirs ) {
+	return generateExtensionsModuleString( getReducersSync( extensionDirs ) );
+}
+
+function getReducersAsync( extensionDirs = [] ) {
+	return new Promise( ( resolveAll ) => {
+		Promise.all(
+			extensionDirs.map( ( extensionDir ) => {
+				return new Promise( ( resolveOne ) => {
+					fs.access( getReducerPath( extensionDir ), fs.constants.F_OK, ( err ) => {
+						resolveOne( ( ! err ) && generateReducerRequireString( extensionDir ) );
+					} );
+				} );
+			} )
+		).then( ( reducerRequires ) => resolveAll( compact( reducerRequires ) ) );
+	} );
+}
```

Comment:
Again, not a blocker, but I'm realizing that

```js
new Promise( resolveAll ) =>
  Promise.all( promises ).then( requires => resolveAll( compact( requires ) ) )
```

is probably equivalent to

```js
Promise.all( promises ).then( compact )
```

Worth looking into after this gets merged — less logic means easier to maintain code, and it's easier to refactor when it's still fresh in our minds.